hnix-store-core/src/System/Nix/Nar.hs

```diff
@@ -0,0 +1,169 @@
+module System.Nix.Nar where
+
+import           Control.Applicative
+import           Control.Monad              (replicateM, replicateM_)
+import qualified Data.Binary                as B
+import qualified Data.Binary.Get            as B
+import qualified Data.Binary.Put            as B
+import qualified Data.ByteString.Lazy.Char8 as BSL
+import           Data.Maybe                 (fromMaybe)
+import           Data.Monoid                ((<>))
+import qualified Data.Set                   as Set
+import qualified Data.Text                  as T
+import qualified Data.Text.Encoding         as E
+import           GHC.Int                    (Int64)
+
+import           System.Nix.Path
+
+
+data NarEffects (m :: * -> *) = NarEffets {
+    readFile         :: FilePath -> m BSL.ByteString
+  , listDir          :: FilePath -> m [FileSystemObject]
+  , narFromFileBytes :: BSL.ByteString -> m Nar
+  , narFromDirectory :: FilePath -> m Nar
+}
+
+
+-- Directly taken from Eelco thesis
+-- https://nixos.org/%7Eeelco/pubs/phd-thesis.pdf
+
+-- TODO: Should we use rootedPath, validPath rather than FilePath?
+data Nar = Nar { narFile :: FileSystemObject }
+    deriving (Eq, Ord, Show)
+
+data FileSystemObject =
+    Regular IsExecutable BSL.ByteString
+  | Directory (Set.Set (PathName, FileSystemObject))
+  | SymLink BSL.ByteString
+  deriving (Eq, Show)
+
+-- TODO - is this right? How does thesis define ordering of FSOs?
+instance Ord FileSystemObject where
+    compare (Regular _ c1) (Regular _ c2) = compare c1 c2
+    compare (Regular _ _)  _              = GT
+    compare (Directory s1) (Directory s2) = compare s1 s2
+    compare (Directory _)  _              = GT
+    compare (SymLink l1) (SymLink l2)     = compare l1 l2
+
+data IsExecutable = NonExecutable | Executable
+    deriving (Eq, Show)
+
+
+------------------------------------------------------------------------------
+-- | Serialize Nar to lazy ByteString
+putNar :: Nar -> B.Put
+putNar (Nar file) = header <>
+                    parens (putFile file)
+    where
+
+        header   = str "nix-archive-1"
+
+        putFile (Regular isExec contents) =
+               str "type" >> str "regular"
+            >> (if isExec == Executable
+               then str "executable" <> str ""
+               else return ())
+            >> str "contents" <> str contents
+
+        putFile (SymLink target) =
+               str "type" >> str "symlink" >> str "target" >> str target
+
+        putFile (Directory entries) =
+               str "type" <> str "directory"
+            <> foldMap putEntry entries
+
+        putEntry (PathName name, fso) =
+            str "entry" <>
+            parens (str "name" <>
+                    str (BSL.fromStrict $ E.encodeUtf8 name) <>
+                    str "node" <>
+                    parens (putFile fso))
+
+        parens m = str "(" >> m >> str ")"
+
+        str :: BSL.ByteString -> B.Put
+        str t = let len = BSL.length t
+            in int len <> pad t
```

Comment:
Hmm... I'm wondering if lazy IO is insufficient here, if we need to send the len before we send the body. Seems like we'll run into the same issues the C++ code has, where it has to dump the entire thing into memory before it can do anything useful. We might instead need to represent a file as a size + a bytestring, where the size is determined from fstat or whatever, and if the bytestring turns out not to be long enough or too long error out.

Reply:
Great point! The Haskell [tar](https://hackage.haskell.org/package/tar) package claims to do laziness correctly - I'll have to read through the code to see if I can find any tips. `BS.length` will definitely force the whole bytestring, so I like the fstat (or whatever :) ) idea.